javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -0,0 +1,416 @@
+
+    override RouteHandler getRouteHandler() { result = this.getRouteHandler() }
+  }
+
+  /**
+   * An invocation that sets any number of headers of the HTTP response.
+   */
+  private class MultipleHeaderDefinitions extends Http::ExplicitHeaderDefinition, DataFlow::CallNode {
+    ReplySource reply;
+
+    MultipleHeaderDefinitions() {
+      // reply.header(RESPONSE, {'Cache-Control': 'no-cache'})
+      // reply(RESPONSE, {'Cache-Control': 'no-cache'})
+      reply.ref().(DataFlow::CallNode).getCalleeName() = ["header", "reply"] and
+      reply.ref().(DataFlow::CallNode).getAnArgument().getALocalSource() instanceof
+        DataFlow::ObjectLiteralNode and
+      this = reply
+    }
+
+    /**
+     * Gets a reference to the multiple headers object that is to be set.
+     */
+    private DataFlow::ObjectLiteralNode getAHeaderSource() {
+      result = this.getAnArgument().getALocalSource()
+    }
+
+    override predicate definesHeaderValue(string headerName, DataFlow::Node headerValue) {
+      exists(string header |
+        this.getAHeaderSource().hasPropertyWrite(header, headerValue) and
+        headerName = header.toLowerCase()
+      )
+    }
+
+    override DataFlow::Node getNameNode() {
+      result = this.getAHeaderSource().getAPropertyWrite().getPropertyNameExpr().flow()
+    }
+
+    override RouteHandler getRouteHandler() { result = reply.getRouteHandler() }
+  }
+
+  /**
+   * A header produced by a route handler with no explicit declaration of a Content-Type.
+   */
+  private class ContentTypeRouteHandlerHeader extends Http::ImplicitHeaderDefinition,
+    DataFlow::FunctionNode instanceof RouteHandler {
+    override predicate defines(string headerName, string headerValue) {
+      headerName = "content-type" and headerValue = "application/json"
+    }
+
+    override Http::RouteHandler getRouteHandler() { result = this }
+  }
+
+  /**
+   * An HTTP cookie defined in a Spife HTTP response.
+   */
+  private class CookieDefinition extends Http::CookieDefinition, DataFlow::MethodCallNode instanceof ReplySource {
+    CookieDefinition() {
+      // reply.cookie(RESPONSE, 'TEST', 'FOO', {"maxAge": 1000, "httpOnly": true, "secure": true})
+      this.ref().(DataFlow::MethodCallNode).getMethodName() = "cookie"
+    }
+
+    override DataFlow::Node getNameArgument() { result = this.getArgument(1) }
+
+    override DataFlow::Node getValueArgument() { result = this.getArgument(2) }
+
+    override RouteHandler getRouteHandler() { result = this.getRouteHandler() }
+  }
+
+  /**
+   * A response argument passed to the `reply` method.
+   */
+  private class ReplyArgument extends Http::ResponseSendArgument, DataFlow::Node {
+    RouteHandler rh;
+
+    ReplyArgument() {
+      exists(ReplySource reply, DataFlow::CallNode call |
+        reply.ref() = call and
+        call.getCalleeName() =
+          ["reply", "cookie", "link", "header", "headers", "raw", "status", "toStream", "vary"] and
```

Comment:
I think this is the same problem as some of the above.  

Here you are asking for a reference to `reply` that is a call to `cookie`, `link`, etc.   
What you want is a method called on a reference to `reply` named `cookie`, `link`, etc.

Reply:
You can try to delete `, "cookie", "link", "header", "headers", "raw", "status", "toStream", "vary"`, and see that all tests still pass (I think that should happen).

Reply:
The problem is that you can use the reply variable directly or a method on it:

```
const reply = require("...")

reply(RESPONSE, {'Cache-Control': 'no-cache'})
reply.header(RESPONSE, {'Cache-Control': 'no-cache'})
```

So I modelled `ReplySource` as a `DataFlow::CallNode` and `reply.ref()` should be a `CallNode` too. Not sure if this is the best way to model it but it worked.

> What you want is a method called on a reference to reply named cookie, link, etc.

That is true for the `reply.header(RESPONSE, {'Cache-Control': 'no-cache'})` case, but not for the `reply(RESPONSE, {'Cache-Control': 'no-cache'})` one, right?

Reply:
I ended up using the same approach that you proposed for the `MultipleHeaderDefinitions` predicate:

```ql
    ReplyArgument() {
      exists(ReplySource reply, DataFlow::CallNode call |
        call =
          [
            // reply(RESPONSE, {'Cache-Control': 'no-cache'})
            reply.ref(),
            // reply.header(RESPONSE, {'Cache-Control': 'no-cache'})
            reply
                .ref()
                .getAMethodCall([
                    "cookie", "link", "header", "headers", "raw", "status", "toStream", "vary"
                  ])
          ]
      |
        this = call.getArgument(0) and
        rh = reply.getRouteHandler()
      )
      or
      this = rh.getAReturn()
    }
```